content/blog/enforcing-and-automatically-tagging-aws-resources/index.md

```diff
@@ -0,0 +1,745 @@
+            Tags: pulumi.Map({
+                "User:Project": pulumi.String(ctx.Project()),
+                "User:Stack": pulumi.String(ctx.Stack()),
+                "User:Cost Center": pulumi.String(config.Require(ctx, "costCenter")),
+            })
+        })
+        return err
+    }
+}
+```
+
+```csharp
+using Pulumi;
+using Pulumi.Aws.S3;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class Program {
+    static Task Main() {
+        return Deployment.RunAsync(() => {
+            // Create an S3 Bucket (with tags):
+            var config = new Config();
+            var bucket = new Bucket("my-bucket", new BucketArgs {
+                Tags: new Dictionary<string, string> {
+                    { "User:Project",     Deployment.Instance.ProjectName },
+                    { "User:Stack",       Deployment.Instance.StackName },
+                    { "User:Cost Center", config.Get("costCenter") },
+                },
+            });
+        });
+    }
+}
+```
+
+In this example, we're using the project and stack names from the current project, and requiring an explicit cost center configured using `pulumi config set costCenter 11223344`. Of course, this is just for illustration purposes &mdash; we could easily used anything for these keys and values.
+
+Suppose our team requires certain tags, though: How do we ensure we don't forget them?
+
+## Enforcing Tagging Strategies
+
+Policy as Code is a way to enforce infrastructure policies, such as ensuring we're not opening databases to the internet, that we're leveraging strong encryption in all the right places, that we aren't violating cost policies, and so on. And it's a great way to ensure we're tagging everything.
+
+Imagine we forgot to tag our S3 Bucket from earlier:
+
+{{< langchoose csharp >}}
+
+```javascript
+// Oops -- no tags!
+let bucket = new aws.s3.Bucket("my-bucket");
+```
+
+```typescript
+// Oops -- no tags!
+const bucket = new aws.s3.Bucket("my-bucket");
+```
+
+```python
+# Oops -- no tags!
+bucket = aws.s3.Bucket('my-bucket')
+```
+
+```go
+// Oops -- no tags!
+_, err := s3.NewBucket(ctx, "my-bucket", nil)
+return err
+```
+
+```csharp
+// Oops -- no tags!
+var bucket = new Bucket("my-bucket");
+```
+
+By applying the policy defined below, deployments that are missing the required tags will fail:
+
+![Tag Policy Failed](./tag-policy-failed.png)
+
+This approach not only ensures we don't forget, but if you're an infrastructure or security engineer, you can now apply this across your team to make sure your team doesn't forget either.
+
+### Defining our Tags Enforcer Policy
+
+This policy is ultimately defined as a simple policy pack as follows:
+
+<!--
+    NOTE: we repeat for languages that aren't TypeScript, because otherwise they'd be
+    hidden based on the above language chooser selection. We are close to being able to
+    have a Python example, however it doesn't yet support configuration.
+-->
```

Comment:
@cnunciato, the new chooser could be used here, which would avoid having to do this, right? The new chooser could be used to specify just JS and TS, and to not hide when other choosers on the page select other languages like Python or Go.

Reply:
@joeduffy, Christian has a great writeup here: https://github.com/pulumi/docs/tree/master/components#pulumi-chooser

Reply:
Untested, but I think you can structure this as follows, using the "local" mode

````markdown
{{< chooser language "javascript,typescript" local >}}

{{% choosable language javascript %}}

```javascript
// the javascript here
```

{{% /choosable %}}
{{% choosable language typescript %}}

```typescript
// the javascript here
```

{{% /choosable %}}

{{< /chooser >}}
````

Reply:
_Nice_!! I was actually going to ask you if there was a better way to do this 😄

Reply:
Yep, it should work exactly like this -- just note that you can't mix `langchoose` and `chooser` shortcodes on the same page, because the existing chooser.js code will interfere by `$.hide()`ing some code blocks. So if you use a `chooser` for one example, you'll need to use `chooser`s for all of them. But other than that, it should do what you expect!